signer/local_cosigner.go

```diff
@@ -69,92 +31,162 @@ type LocalCosignerConfig struct {
-	rsaKey      rsa.PrivateKey
-	total       uint8
-	threshold   uint8
-
-	// stores the last sign state for a share we have fully signed
-	// incremented whenever we are asked to sign a share
-	lastSignState *SignState
-
-	// signing is thread safe
-	lastSignStateMutex sync.Mutex
-
-	// Height, Round, Step -> metadata
-	hrsMeta map[HRSTKey]HrsMetadata
-	peers   map[int]CosignerPeer
-
-	address string
-}
-
-func (cosigner *LocalCosigner) SaveLastSignedState(signState SignStateConsensus) error {
-	return cosigner.lastSignState.Save(signState, &cosigner.lastSignStateMutex)
+	/*
+		// key         CosignerKey		- moved to threshold
+		// rsaKey      rsa.PrivateKey	- moved to threshold
+		// threshold uint8 				- moved to threshold
+
+		// Height, Round, Step -> metadata 	- moved to threshold
+		//hrsMeta map[HRSTKey]HrsMetadata 	- moved to threshold
+
+	*/
+	LastSignStateStruct *LastSignStateStruct
+	pubKeyBytes         []byte
+	total               uint8
+	address             string
+	// peers       map[int]CosignerPeer
+	localsigner *LocalSoftSignThresholdEd25519Signature // TODO: Change the mutex so that you can you can pass ThresholdEd25519Signature
 }
 
 func NewLocalCosigner(cfg LocalCosignerConfig) *LocalCosigner {
+	// TODO: localsigner should be passed as a parameter in the cfg rather than constructed here. We could add a method in the new LocalSoftSignThresholdEd25519Signature file to return this config such as func NewLocalThresholdEd25519Signature(key CosignerKey, rsaKey PrivateKey, total, threshold int) LocalSoftSignThresholdEd25519Signature { which could then be called and added to the LocalCosignerConfig from cmd/horcrux/cmd/cosigner.go
+
+	localsigner := LocalSoftSignThresholdEd25519Signature{
+		Key:       cfg.CosignerKey,
+		RsaKey:    cfg.RsaKey,
+		HrsMeta:   make(map[HRSTKey]HrsMetadata),
+		Peers:     make(map[int]CosignerPeer),
+		Total:     cfg.Total,
+		Threshold: cfg.Threshold,
+	}
+
+	LastSignStateStruct := LastSignStateStruct{
+		LastSignStateMutex: sync.Mutex{},
+		LastSignState:      cfg.SignState,
+	}
 	cosigner := &LocalCosigner{
-		key:           cfg.CosignerKey,
-		lastSignState: cfg.SignState,
-		rsaKey:        cfg.RsaKey,
-		hrsMeta:       make(map[HRSTKey]HrsMetadata),
-		peers:         make(map[int]CosignerPeer),
-		total:         cfg.Total,
-		threshold:     cfg.Threshold,
-		address:       cfg.Address,
+		//key:         cfg.CosignerKey,
+		//pubKeyBytes: []byte{},
+		LastSignStateStruct: &LastSignStateStruct,
+		total:               cfg.Total,
+		address:             cfg.Address,
+		localsigner:         &localsigner,
 	}
 
+	// TODO: Delete this print statements:
+	// fmt.Println("\n", "LocalCosigner")
+	// fmt.Printf("%+v\n", cosigner)
+
+	// fmt.Println("\n", "Local Signer")
+	// fmt.Printf("%+v\n", cosigner.localsigner)
+
 	for _, peer := range cfg.Peers {
-		cosigner.peers[peer.ID] = peer
+		cosigner.localsigner.Peers[peer.ID] = peer
 	}
 
 	// cache the public key bytes for signing operations
```

Comment:
This block of code can move into a constructor method for the `LocalSoftSignThresholdEd25519Signature`, e.g. `NewLocalThresholdEd25519Signature`

Reply:
Yes indeed, haven't started on factorising the constructor for NewLocalThresholdEd25519Signature/LocalSoftSignThresholdEd25519Signature

Reply:
This should stay put in the local_cosigner constructor since we have moved back the Peers up to the local_cosigner level.